tools/cache/time-defined-cache.go

```diff
@@ -0,0 +1,97 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright The KubeVirt Authors
+ *
+ */
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type TimeDefinedCache[T any] struct {
+	minRefreshDuration time.Duration
+	lastRefresh        time.Time
+	savedValueSet      bool
+	savedValue         T
+	reCalcFunc         func() (T, error)
+	valueLock          *sync.RWMutex
+}
+
+// NewTimeDefinedCache creates a new cache that will refresh the value every minRefreshDuration. If the value is requested
+// before the minRefreshDuration has passed, the cached value will be returned. If minRefreshDuration is zero, the value will always be
+// recalculated.
+// In addition, a Set() can be used to explicitly set the value.
+// If useValueLock is set to true, the value will be locked when being set. If the cache won't be used concurrently, it's safe
+// to set this to false.
+func NewTimeDefinedCache[T any](minRefreshDuration time.Duration, useValueLock bool, reCalcFunc func() (T, error)) (*TimeDefinedCache[T], error) {
+	if reCalcFunc == nil {
+		return nil, fmt.Errorf("re-calculation function is not set")
+	}
+
+	t := &TimeDefinedCache[T]{
+		minRefreshDuration: minRefreshDuration,
+		reCalcFunc:         reCalcFunc,
+	}
+
+	if useValueLock {
+		t.valueLock = &sync.RWMutex{}
+	}
+
+	return t, nil
+}
+
+func (t *TimeDefinedCache[T]) Get() (T, error) {
+	if t.savedValueSet && t.minRefreshDuration.Nanoseconds() != 0 && time.Since(t.lastRefresh) <= t.minRefreshDuration {
+		if t.valueLock != nil {
+			t.valueLock.RLock()
+			defer t.valueLock.RUnlock()
+		}
+		return t.savedValue, nil
+	}
```

Comment:
Hey @vasiliy-ul! Thanks for having a look!

> The lock should also protect t.savedValueSet and t.lastRefresh fields. Otherwise, there is a potential race with set/recalc flow.

I'm not sure I see the potential danger here. Would you be willing to elaborate on the exact flow that might be dangerous? Something similar to what I did [here](https://github.com/kubevirt/kubevirt/pull/11634#discussion_r1551414254) would be great. Thanks!

> Also, a small nit: instead of having a separate field savedValueSet, you could store savedValue as a pointer and check for nil.

Sounds good to me!

Reply:
> I'm not sure I see the potential danger here. Would you be willing to elaborate on the exact flow that might be dangerous?

There is no "danger" in the sense of data corruption or undefined behaviour, since it is a **read** section. However, this may return stale data or trigger unnecessary recalc, e.g.:

* thread1 calls `Get` when `t.savedValueSet` is false (i.e. first time call). Then it calls `t.reCalcFunc()`
* thread2 gets in and also calls 'Get`. `t.savedValueSet` may still be not set and since it is not protected by read lock, the thread will also do the recalc.

Reply:
@vasiliy-ul So thinking about this again, even with your suggestion the same race can occur:
* thread1 calls the anonymous function, results with `ok == false`
* thread2 calls the anonymous function, results with `ok == false`
* thread 1 captures the write lock
* thread 2 waits on the write lock
* thread 1 calculates value and sets it
* thread 2 calculates value and sets it

I'm not sure how we can get around it without giving up on the RWMutex and going back to a regular mutex (or IOW, always capture the write lock until `Get()` as a whole is done.

I think this option is actually reasonable. I would guess that the performance differences are negligible, and perhaps it's more important to avoid such races that would end up in re-calculation and much more severe performance penalties.

WDYT? @vasiliy-ul @fossedihelm

Reply:
@iholder101, yeah, you are right. Probably protecting the whole `Get()` would make sense then. Especially since it is not a pure **read** but also potentially a **write**. I think a regular lock is indeed more appropriate here.

Reply:
Nice catch @vasiliy-ul! @iholder101 Agree! This `Get` can perform a `write` so I think that regular lock is a better solution, and maybe the only one which works.
Thank you!

Reply:
@vasiliy-ul TBH, I'm not sure it worths the effort and the code being more messy. I tend to think that the performance difference is negligible and we're over-thinking about micro-optimizations here. I think locking `Get()` as a whole is good enough TBH.

Reply:
Sure. Each optimization needs to be justified first, and here we do not have any performance data. I am totally fine with the simple approach. Just wanted to have this option here for the record.

Reply:
Thanks a lot @vasiliy-ul and @fossedihelm. Your review is very valuable.

I've opened a follow-up here: https://github.com/kubevirt/kubevirt/pull/11663